ccl/backupccl: TestRestoreAsOfSystemTimeGCBounds failed

ccl/backupccl.TestRestoreAsOfSystemTimeGCBounds [failed](https://tanzanite.cluster.engflow.com/invocations/default/9566c8fa-20be-4419-bb3d-d7f43e01ec0b?testReportRun=1&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvY2NsL2JhY2t1cGNjbDpiYWNrdXBjY2xfdGVzdA==) on release-24.1 @ [bebd9a7d30104773ff2b85d8af0785a1b2f00fab](https://github.com/cockroachdb/cockroach/commits/bebd9a7d30104773ff2b85d8af0785a1b2f00fab):

Fatal error:

```
panic: test timed out after 1h16m37s
running tests:
	TestRestoreAsOfSystemTimeGCBounds (1h16m37s)

```
Stack: 

```
goroutine 728404 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x265
created by time.goFunc
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:471 +0x69d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
	GOROOT/src/sync/cond.go:70 +0x75
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*schedulerShard).processEvents(0xc004321680, {0xe2d0558, 0xc001b645d0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/scheduler.go:482 +0x79c
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*Scheduler).Start.func1({0xe2d0558, 0xc001b645d0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/scheduler.go:255 +0x1f5
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:480 +0x263
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 150
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:471 +0x69d

goroutine 154 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc00431c950, 0x3176)
	GOROOT/src/runtime/sema.go:569 +0x159
sync.(*Cond).Wait(0xc00431c940)
	GOROOT/src/sync/cond.go:70 +0x75
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*schedulerShard).processEvents(0xc004321680, {0xe2d0558, 0xc001b64660})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/scheduler.go:482 +0x79c
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*Scheduler).Start.func1({0xe2d0558, 0xc001b64660})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/scheduler.go:255 +0x1f5
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:480 +0x263
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 150
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:471 +0x69d

goroutine 155 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc00431c950, 0x3172)
	GOROOT/src/runtime/sema.go:569 +0x159
sync.(*Cond).Wait(0xc00431c940)
	GOROOT/src/sync/cond.go:70 +0x75
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*schedulerShard).processEvents(0xc004321680, {0xe2d0558, 0xc001b646f0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/scheduler.go:482 +0x79c
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*Scheduler).Start.func1({0xe2d0558, 0xc001b646f0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/scheduler.go:255 +0x1f5
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:480 +0x263
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 150
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:471 +0x69d

goroutine 119 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc00431c950, 0x3175)
	GOROOT/src/runtime/sema.go:569 +0x159
sync.(*Cond).Wait(0xc00431c940)
	GOROOT/src/sync/cond.go:70 +0x75
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*schedulerShard).processEvents(0xc004321680, {0xe2d0558, 0xc001b47050})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/scheduler.go:482 +0x79c
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*Scheduler).Start.func1({0xe2d0558, 0xc001b47050})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/scheduler.go:255 +0x1f5
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:480 +0x263
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 150
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-50586

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -2052,7 +2052,7 @@ func (m *Manager) refreshSomeLeases(ctx context.Context, refreshAndPurgeAllDescr
 				if _, err := acquireNodeLease(ctx, m, id, AcquireBackground); err != nil {
 					log.Errorf(ctx, "refreshing descriptor: %d lease failed: %s", id, err)
 
-					if errors.Is(err, catalog.ErrDescriptorNotFound) {
+					if errors.Is(err, catalog.ErrDescriptorNotFound) || errors.Is(err, catalog.ErrDescriptorDropped) {
 						// Lease renewal failed due to removed descriptor; Remove this descriptor from cache.
 						if err := purgeOldVersions(
 							ctx, m.storage.db.KV(), id, true /* dropped */, 0 /* minVersion */, m,

--- a/pkg/sql/catalog/lease/lease_test.go
+++ b/pkg/sql/catalog/lease/lease_test.go
@@ -3213,9 +3213,17 @@ func TestLeaseDescriptorRangeFeedFailure(t *testing.T) {
 							p.Params.ExecutionPhase != scop.PostCommitPhase {
 							return nil
 						}
+						foundDescriptorDrop := false
+						// Ensure we have a descriptor drop that will be executed before
+						// we mess with the range feed.
+						for _, op := range p.Stages[stageIdx].EdgeOps {
+							if _, ok := op.(*scop.MarkDescriptorAsDropped); ok {
+								foundDescriptorDrop = true
+							}
+						}
 						// Once this stage completes, we can "resume" the range feed,
 						// so the update is detected.
-						if stageIdx == 0 {
+						if foundDescriptorDrop {
 							rangeFeedResetChan = srv.ApplicationLayer(1).LeaseManager().(*lease.Manager).TestingSetDisableRangeFeedCheckpointFn(true)
 							enableAfterStageKnob.Swap(false)
 							grp.Go(func() error {
@@ -3239,15 +3247,20 @@ func TestLeaseDescriptorRangeFeedFailure(t *testing.T) {
 	// detects a problem.
 	defer srv.ApplicationLayer(1).LeaseManager().(*lease.Manager).TestingSetDisableRangeFeedCheckpointFn(false)
 	firstConn.Exec(t, "CREATE TABLE t1(n int)")
+	firstConn.Exec(t, "CREATE TABLE t2(n int)")
 	require.NoError(t, srv.WaitForFullReplication())
 	tx := secondConn.Begin(t)
 	_, err := tx.Exec("SELECT * FROM t1;")
+	require.NoError(t, err)
+	_, err = tx.Exec("SELECT * FROM t2;")
 	require.NoError(t, err)
 	// This schema change will wait for the connection on
 	// node 1 to release the lease. Because the rangefeed is
 	// disabled it will never know about the new version.
 	enableAfterStageKnob.Store(true)
-	firstConn.Exec(t, "ALTER TABLE t1 ADD COLUMN j INT DEFAULT 64")
+	firstConn.Exec(t, "SET autocommit_before_ddl=false")
+	firstConn.Exec(t, "SET use_declarative_schema_changer='unsafe_always'")
+	firstConn.Exec(t, "ALTER TABLE t1 ADD COLUMN j INT DEFAULT 64; DROP TABLE t2;")
 	_, err = tx.Exec("INSERT INTO t1 VALUES (32)")
 	if err != nil {
 		t.Fatal(err)

